### PR TITLE
Automate DevConsole Interop WTO Scenarios

### PR DESCRIPTION
**Pull request by musienko-maxim**
**Jira Id:** https://issues.redhat.com/browse/WTO-73
**Checks for approving Epic scenarios Automation PR:**

- [ ] Execute the @to-do tagged gherkin scripts manually
- [x] Convert the @to-do gherkin scripts to cypress automation scripts
- [x] Once scripts are automated, replace tag @to-do with @epic-number
- [x] Execute the scripts in Remote cluster
**Cluster Setup:**
```
- export NO_HEADLESS=true && export CHROME_VERSION=$(/usr/bin/google-chrome-stable --version)
- BRIDGE_KUBEADMIN_PASSWORD=******
- BRIDGE_BASE_ADDRESS=https://console-openshift-console.apps.dev-svc-4.8-042801.devcluster.openshift.com
- export BRIDGE_KUBEADMIN_PASSWORD`
- export BRIDGE_BASE_ADDRESS
- oc login -u kubeadmin -p $BRIDGE_KUBEADMIN_PASSWORD
- oc apply -f ./frontend/integration-tests/data/htpasswd-secret.yaml
- oc patch oauths cluster --patch "$(cat ./frontend/integration-tests/data/patch-htpasswd.yaml)" --type=merge
```
***Test Setup:***
* Replace "TAGS": "@smoke and not @manual and not @to-do", with " (@smoke or @regression) and not (@Manual or @to-do)", in topology/cypress.json under env
* Run ./test-cypress.sh -p topology in console
* Execute feature files from: frontend/packages/dev-console/integration-tests/features/web-terminal folder
1. web-teminal-basic.feature
2. web-terminal-adminuser.feature
3. web-terminal-devuser.feature

**Results**:
Selection_012
Selection_013
Selection_014

### CODE REVIEW

File: frontend/packages/dev-console/integration-tests/support/step-definitions/web-terminal/web-teminal-basic.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import { When, Then, And } from 'cypress-cucumber-preprocessor/steps';
+import { webTerminalPage } from '../../pages/web-terminal/webTerminal-page';
+
+When('user clicks on Open Terminal in new tab button on the terminal window', () => {
+  webTerminalPage.verifyOpenInNewTabButton();
+});
+And('user close current Web Terminal session', () => {
```
Comment (gajanan-more) [inline frontend/packages/dev-console/integration-tests/support/step-definitions/web-terminal/web-teminal-basic.ts:7]: Can you please replace `And` with either `Given`, `When` or `Then`?

File: frontend/packages/dev-console/integration-tests/testData/yamls/web-terminal/web-terminal.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    # it's a label OpenShift console uses a flag to mark terminal's workspaces
+    console.openshift.io/terminal: "true"
+spec:
+  started: true
+  routingClass: 'web-terminal'
+  template:
+    components:
+      - plugin:
+          name: web-terminal
+          id: redhat-developer/web-terminal/latest
```
Comment (gajanan-more) [inline frontend/packages/dev-console/integration-tests/testData/yamls/web-terminal/web-terminal.yaml:19]: Can you please add new empty line here?

File: frontend/packages/dev-console/integration-tests/support/step-definitions/web-terminal/web-Terminal-devuser.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import { When, Then, Given } from 'cypress-cucumber-preprocessor/steps';
+import { initTerminalPage } from '../../pages/web-terminal/initTerminal-page';
+import { searchResource } from '../../pages/search-resources/search-page';
+import { devWorkspacePage } from '../../pages/devworspace/devworkspacePage';
+import { devWorkspaceStatuses, switchPerspective } from '../../constants';
+import { projectNameSpace, perspective } from '../../pages';
+
```
Comment (gajanan-more) [inline frontend/packages/dev-console/integration-tests/support/step-definitions/web-terminal/web-Terminal-devuser.ts:7]: Can you please remove this empty line?

**Review by gajanan-more — CHANGES_REQUESTED** [review]

**Review by gajanan-more — COMMENTED** [review]
Please add a proper description to this PR. Share the test results screenshot as well. Please refer this [PR](https://github.com/openshift/console/pull/9177) or this example [template](https://github.com/openshift/console/blob/master/frontend/packages/dev-console/integration-tests/.github/PULL_REQUEST_TEMPLATE/qe-automation-pr.md)

File: frontend/packages/dev-console/integration-tests/support/step-definitions/web-terminal/web-Terminal-devuser.ts (hunk 3 above)
Comment (musienko-maxim) [reply frontend/packages/dev-console/integration-tests/support/step-definitions/web-terminal/web-Terminal-devuser.ts]: Remove unnecessary line

File: frontend/packages/dev-console/integration-tests/testData/yamls/web-terminal/web-terminal.yaml (hunk 2 above)
Comment (musienko-maxim) [reply frontend/packages/dev-console/integration-tests/testData/yamls/web-terminal/web-terminal.yaml:19]: Add empty line

File: frontend/packages/dev-console/integration-tests/support/step-definitions/web-terminal/web-teminal-basic.ts (hunk 1 above)
Comment (musienko-maxim) [reply frontend/packages/dev-console/integration-tests/support/step-definitions/web-terminal/web-teminal-basic.ts:7]: Replace on Then

**Review by gajanan-more — CHANGES_REQUESTED** [review]
Please update the PR description. It still says to update the `cypress.json` from the topology plugin. 
Please remove the `@to-do` tags. If the scenarios have a `@to-do` tag, the automation script will skip those scenarios.

**Comment by gajanan-more** [comment]
Screenshot from 2021-07-19 17-01-31

@makambalaji  test cases are failing because of the above issue. Cna you please have a look?

**Comment by makambalaji** [comment]
> Screenshot from 2021-07-19 17-01-31
> 
> @makambalaji test cases are failing because of the above issue. Cna you please have a look?

[consoledeveloper.sh](https://github.com/redhat-developer/devconsole-operator/blob/master/hack/install_devconsole/consoledeveloper.sh) needs to be executed before these scripts, because clusters created usually doesn't contain **console developer** user
@musienko-maxim - Can you please add this step

**Comment by makambalaji** [comment]
Feature files looks good and there is no much modifications to affect the existing code. so adding lgtm. But scripts are failing. need to raise a separate bug and should visit again

/lgtm

**Comment by makambalaji** [comment]
Adding below labels for merging purpose.
/label px-approved
/label qe-approved
/label docs-approved

**Comment by makambalaji** [comment]
/hold
PR contains merge conflict, needs to fix it

File: frontend/packages/dev-console/integration-tests/support/step-definitions/common/common.ts [hunk 4]
```diff
@@ -35,7 +35,7 @@ Given('user has created namespace starts with {string}', (projectName: string) =
 
 Given('user has created or selected namespace {string}', (projectName: string) => {
   Cypress.env('NAMESPACE', projectName);
-  projectNameSpace.selectOrCreateProject(`${projectName}`);
+  projectNameSpace.selectOrCreateProjectForVersion49(`${projectName}`);
```
Comment (jrichter1) [inline frontend/packages/dev-console/integration-tests/support/step-definitions/common/common.ts:38]: There seems to be no reason to change a step common for all the tests. I've tried running the tests with the original `selectOrCreateProject` and they seem to be running just as well.

Comment (musienko-maxim) [reply frontend/packages/dev-console/integration-tests/support/step-definitions/common/common.ts:38]: Strange... Let me recheck it again.

Comment (musienko-maxim) [reply frontend/packages/dev-console/integration-tests/support/step-definitions/common/common.ts:38]: Maybe I  something lost but when I Opened the Widget: 
Selection_024
I did not find the WebElement like: https://github.com/openshift/console/blob/master/frontend/packages/dev-console/integration-tests/support/pages/app.ts#L174

Comment (musienko-maxim) [reply frontend/packages/dev-console/integration-tests/support/step-definitions/common/common.ts:38]: Looks like it works as expected under the admin perspective but if we login under regular user it fails, we need to modify this function for both cases.

**Comment by psrna** [comment]
@musienko-maxim Hi Maksym, any updates on this? Thx

**Comment by musienko-maxim** [comment]
reopen

**Review by ibuziuk — APPROVED** [review]

**Comment by sanketpathak** [comment]
@musienko-maxim Getting this when running web-terminal-basic.feature file
Screenshot from 2022-08-25 12-12-24

**Comment by musienko-maxim** [comment]
> @musienko-maxim Getting this when running web-terminal-basic.feature file Screenshot from 2022-08-25 12-12-24

I've not fixed the `web-terminal-devuser.feature` and `web-teminal-basic.feature`. I do not know which option will be better to fix it in this commit or merge it and create a second one?

File: frontend/packages/dev-console/integration-tests/features/web-terminal/web-teminal-basic.feature [hunk 5]
```diff
@@ -3,29 +3,28 @@ Feature: Web Terminal
...
         Scenario: Web Terminal in new tab: WT-01-TC02
             Given user can see terminal icon on masthead
              When user clicks on the Web Terminal icon on the Masthead
               And user clicks on Open Terminal in new tab button on the terminal window
              Then user will see the terminal window opened in new tab
 
-
         @regression @to-do
-        Scenario: Web Terminal is stopped by inactivity: WT-01-TC03
+        Scenario: user does nothing with displayed terminal window 1 minutes: WT-01-TC03
```
Comment (jerolimov) [inline frontend/packages/dev-console/integration-tests/features/web-terminal/web-teminal-basic.feature:25]: Hey, the new scenario title says what the user does, not what is expected. I would recommend to keep the old title for now or is there any reason to change this title?

```suggestion
        Scenario: Web Terminal is stopped by inactivity: WT-01-TC03
```

**Comment by sanketpathak** [comment]
Ran web-terminal-adminuser.feature, working fine
Screenshot from 2022-09-01 12-16-48
/lgtm

**Pull request merged**